content/200-concepts/100-components/03-prisma-migrate/010-mental-model.mdx

```diff
@@ -0,0 +1,163 @@
+
+![Prisma Migrate dev flow](https://user-images.githubusercontent.com/33921841/189875756-5df6ba94-c7a9-4e9c-a8f7-ce4d5d09b204.png)
+
+The `prisma migrate dev` command tracks the state of the database using the following pieces of state:
+
+- the Prisma schema
+- the migrations history
+- the migrations table
+- the database schema
+
+> **Note**: The pieces of state used to track the state of a migration are the same as the ones described in [how Prisma Migrate tracks the migration state](#how-prisma-migrate-tracks-the-migration-state) section.
+
+You can customize migrations before you apply them to the database using the `--create-only` flag. For example, you might want to edit a migration if you want to rename columns without incurring any data loss or load database extensions (in PostgreSQL) and database views (currently not supported).
+
+Under the hood, Prisma Migrate uses a [shadow database](/concepts/components/prisma-migrate/shadow-database) to detect a [schema drift](/concepts/components/prisma-migrate/shadow-database#detecting-schema-drift) and generate new migrations.
+
+> **Note**: `prisma migrate dev` is intended to be used only in development with a disposable database.
+
+If `prisma migrate dev` detects a schema drift or a migration history conflict, you will be prompted to reset (drop and recreate your database) your database to sync the migration history and the database schema.
+
+<details>
+
+<summary> Expand to see the shadow database explained using a cartoon</summary>
+
+![A cartoon that shows how the shadow database works.](shadow-database.png)
+
+</details>
+
+#### Resolve schema drifts
+
+A schema drift occurs when the expected database schema is different from what is in the migration history. For example, this can occur when you manually update the database schema without also updating the Prisma schema and `prisma/migrations` accordingly.
+
+For such instances, you can use the [`prisma migrate diff`](/reference/api-reference/command-reference#migrate-diff) command to compare your migration history and revert changes made to your database schema.
+
+![Revert database schema with `migrate diff`](https://user-images.githubusercontent.com/33921841/189894555-b1dedf54-a8eb-4e29-9bf4-85a5615030b4.png)
+
+You can use `migrate diff` to generate the SQL that either:
+
+- Reverts the changes made in the database schema to synchronize it with the current Prisma schema
+- Moves your database schema forward to apply missing changes from the Prisma schema and `/migrations`
+
+You can then apply the changes to your database using [`prisma db execute`](/reference/api-reference/command-reference#db-execute) command.
```

Comment:
`prisma db push` should always be taught after `prisma migrate dev`. My argument for this is the pit of success:
- `prisma migrate dev` is the essential command anyone should learn and understand
- `prisma db push` is extra convenience for a specific use case of `prisma migrate dev` ("prototyping")
- teaching `prisma db push` before `prisma migrate dev` is frontloading complexity of a feature that should be discovered later when needed

```suggestion
### Prisma Migrate in a development environment (local)

#### Track your migration history with `prisma migrate dev`

The [`prisma migrate dev`](/reference/api-reference/command-reference#migrate-dev) command allows you to track the changes you make to your database. The `prisma migrate dev` command automatically generates SQL migration files (saved in `/prisma/migrations`) and applies them to the database. When a migration is applied to the database, the migrations table (`_prisma_migrations`) in your database is also updated.

![Prisma Migrate dev flow](https://user-images.githubusercontent.com/33921841/189875756-5df6ba94-c7a9-4e9c-a8f7-ce4d5d09b204.png)

The `prisma migrate dev` command tracks the state of the database using the following pieces of state:

- the Prisma schema
- the migrations history
- the migrations table
- the database schema

> **Note**: The pieces of state used to track the state of a migration are the same as the ones described in [how Prisma Migrate tracks the migration state](#how-prisma-migrate-tracks-the-migration-state) section.

You can customize migrations before you apply them to the database using the `--create-only` flag. For example, you might want to edit a migration if you want to rename columns without incurring any data loss or load database extensions (in PostgreSQL) and database views (currently not supported).

Under the hood, Prisma Migrate uses a [shadow database](/concepts/components/prisma-migrate/shadow-database) to detect a [schema drift](/concepts/components/prisma-migrate/shadow-database#detecting-schema-drift) and generate new migrations.

> **Note**: `prisma migrate dev` is intended to be used only in development with a disposable database.

If `prisma migrate dev` detects a schema drift or a migration history conflict, you will be prompted to reset (drop and recreate your database) your database to sync the migration history and the database schema.

<details>

<summary> Expand to see the shadow database explained using a cartoon</summary>

![A cartoon that shows how the shadow database works.](shadow-database.png)

</details>

#### Resolve schema drifts

A schema drift occurs when the expected database schema is different from what is in the migration history. For example, this can occur when you manually update the database schema without also updating the Prisma schema and `prisma/migrations` accordingly.

For such instances, you can use the [`prisma migrate diff`](/reference/api-reference/command-reference#migrate-diff) command to compare your migration history and revert changes made to your database schema.

![Revert database schema with `migrate diff`](https://user-images.githubusercontent.com/33921841/189894555-b1dedf54-a8eb-4e29-9bf4-85a5615030b4.png)

You can use `migrate diff` to generate the SQL that either:

- Reverts the changes made in the database schema to synchronize it with the current Prisma schema
- Moves your database schema forward to apply missing changes from the Prisma schema and `/migrations`

You can then apply the changes to your database using [`prisma db execute`](/reference/api-reference/command-reference#db-execute) command.

#### Prototype your schema

The [`prisma db push`](/reference/api-reference/command-reference#db-push) command allows you to sync your Prisma schema and database schema without persisting a migration (`/prisma/migrations`). The `prisma db push` command tracks the state of the database using the following pieces of state:

- the Prisma schema
- the database schema

![prisma db push development flow](https://user-images.githubusercontent.com/33921841/189866483-cec55d83-d1fe-4446-9c28-0a60329f975d.png)

The `prisma db push` command is useful when:

- You want to **quickly prototype and iterate** on schema design locally without the need to deploy these changes to other environments such as other developers, or staging and production environments.
- You are prioritizing reaching a **desired end-state** and not the changes or steps executed to reach that end-state (there is no way to preview changes made by `prisma db push`)
- You do not need to control how schema changes impact data. There is no way to orchestrate schema and data migrations - if `prisma db push` anticipates that changes will result in data loss, you can either accept data loss with the `--accept-data-loss` option or stop the process - there is no way to customize the changes.

If the `prisma db push` command detects a [schema drift](/guides/database/developing-with-prisma-migrate/troubleshooting-development#schema-drift), it will prompt you to reset your database. For example, this will happen when you add a required field to a table with existing content without providing a default value.

> A [schema drift](/guides/database/developing-with-prisma-migrate/troubleshooting-development#schema-drift) occurs when your database schema is out of sync with your migrations history and migrations table.

```